lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+                        _get_collection_info(dependency_map, existing_collections, dep_name, dep_requirement,
+                                             parent_info.source, temp_path, servers, validate_certs, force_deps,
+                                             parent=parent)
+
+                    checked_parents.add(parent)
+
+            # No extra dependencies were resolved, exit loop
+            if deps_exhausted:
+                break
+
+        # Now we have resolved the deps to our best extent, now select the latest version for collections with
+        # multiple versions found and go from there
+        deps_not_checked = set(dependency_map.keys()).difference(checked_parents)
+        for collection in deps_not_checked:
+            dependency_map[collection].set_latest_version()
+            if no_deps or len(dependency_map[collection].dependencies) == 0:
+                checked_parents.add(collection)
+
+    return dependency_map
+
+
+def _get_collection_info(dep_map, existing_collections, collection, requirement, source, temp_path, server_list,
+                         validate_certs, force, parent=None):
+    dep_msg = ""
+    if parent:
+        dep_msg = " - as dependency of %s" % parent
+    display.vvv("Processing requirement collection '%s'%s" % (to_text(collection), dep_msg))
+
+    tar_path = None
+    if os.path.isfile(collection):
+        display.vvvv("Collection requirement '%s' is a tar artifact" % to_text(collection))
+        tar_path = collection
+    elif urlparse(collection).scheme:
+        display.vvvv("Collection requirement '%s' is a URL to a tar artifact" % collection)
+        tar_path = _download_file(collection, temp_path, None, validate_certs)
+
+    if tar_path:
+        req = CollectionRequirement.from_tar(tar_path, validate_certs, force, arent=parent)
+
+        collection_name = str(req)
+        if collection_name in dep_map:
+            collection_info = dep_map[collection_name]
+            collection_info.add_requirement(None, req.latest_version)
+        else:
+            collection_info = req
+    else:
+        display.vvvv("Collection requirement '%s' is the name of a collection" % collection)
+        if collection in dep_map:
+            collection_info = dep_map[collection]
+            collection_info.add_requirement(parent, requirement)
+        else:
+            servers = [source] if source else server_list
+            collection_info = CollectionRequirement.from_name(collection, servers, requirement, validate_certs, force,
+                                                              parent=parent)
+
+    existing = [c for c in existing_collections if str(c) == str(collection_info)]
+    if existing and not collection_info.force:
+        existing[0].add_requirement(parent, requirement)  # Test that the installed collection fits the requirement
+        collection_info = existing[0]
+
+    dep_map[str(collection_info)] = collection_info
+
+
+def _urljoin(*args):
```

Comment:
You do iterable unpacking in all places where you use this. Maybe it's easier to just use accept an iterable rather than a bunch of separate args?
```suggestion
def _urljoin(slugs):
```

Reply:
Doesn't look like that's the case, maybe I've change it in the last commit but I see only 1 instance where I pass a list, the rest are direct args.